File: bench-vs-rust/rust-linebreak/diagnostic.zig
const std = @import("std");
const internal = @import("internal");
const scalar = internal.scalar;
const lb = internal.line_break;
const zunic = @import("zunic");
const cases = .{
    .{ "arabic", @embedFile("texts/arabic.txt") },
    .{ "hindi", @embedFile("texts/hindi.txt") },
    .{ "korean", @embedFile("texts/korean.txt") },
    .{ "russian", @embedFile("texts/russian.txt") },
    .{ "source_code", @embedFile("texts/source_code.txt") },
    .{ "english", @embedFile("texts/english.txt") },
    .{ "japanese", @embedFile("texts/japanese.txt") },
    .{ "mandarin", @embedFile("texts/mandarin.txt") },
};
var tokens: [60000]scalar.ClassifiedToken = undefined;
var token_count: usize = 0;

const Result = struct {
    count: usize = 0,
    checksum: u64 = 0,
    fn add(self: *Result, offset: usize, status: u64) void {
        self.count += 1;
        self.checksum = self.checksum *% 31 +% offset;
        self.checksum = self.checksum *% 31 +% status;
    }
};
fn kind(op: lb.Opportunity) u64 {
    return switch (op) {
        .prohibited => 0,
        .allowed => 1,
        .mandatory => 2,
    };
}

fn classify(bytes: []const u8) Result {
    var input = bytes;
    std.mem.doNotOptimizeAway(&input);
    var classifier = scalar.Classifier(false){};
    var pos: usize = 0;
    var result: Result = .{};
    while (pos < input.len) {
        const token = classifier.at(input, pos);
        result.checksum +%= @intFromEnum(token.record.line_break);
        result.count += 1;
        pos = token.end;
    }
    return result;
}

fn iterate(comptime consume_all: bool, bytes: []const u8) Result {
    var input = bytes;
    std.mem.doNotOptimizeAway(&input);
    var it = lb.iterator(input);
    var result: Result = .{};
    while (it.next()) |boundary| {
        if (consume_all or boundary.opportunity != .prohibited) result.add(boundary.offset, kind(boundary.opportunity));
    }
    return result;
}
fn full(bytes: []const u8) Result {
    return iterate(false, bytes);
}
fn allBoundaries(bytes: []const u8) Result {
    return iterate(true, bytes);
}

fn wrapPass(bytes: []const u8, comptime limit: usize) Result {
    var input = bytes;
    std.mem.doNotOptimizeAway(&input);
    var it = (zunic.text(input).wrap(.{ .max_columns = 80, .overflow = .grapheme }) catch unreachable).iterator();
    var result: Result = .{};
    while (result.count < limit) {
        const line = it.next() orelse break;
        result.add(line.start.value, line.end.value);
        result.checksum = result.checksum *% 31 +% line.columns.value;
    }
    return result;
}
fn wrapFull(bytes: []const u8) Result {
    return wrapPass(bytes, std.math.maxInt(usize));
}
fn wrapViewport(bytes: []const u8) Result {
    return wrapPass(bytes, 24);
}

// Separate diagnostic: predecoded tokens, not an end-to-end Rust comparison.
fn transitions(bytes: []const u8) Result {
    var input = bytes;
    std.mem.doNotOptimizeAway(&input);
    var result: Result = .{};
    if (token_count == 0) {
        result.add(input.len, 2);
        return result;
    }
    const first = tokens[0];
    var state = lb.State.firstWithRecord(first.record.line_break, first.codepoint orelse 0, first.record);
    var classifier = scalar.Classifier(false){};
    const eot = classifier.at(input, input.len);
    for (1..token_count) |i| {
        const token = tokens[i];
        const next = if (i + 1 < token_count) tokens[i + 1] else eot;
        const cp = token.codepoint orelse 0;
        const op = state.opportunityForRecord(input, cp, token.record, next.record.line_break, next.record, i + 1 < token_count, next.end, &classifier);
        state.consumeRecord(cp, token.record);
        if (op != .prohibited) result.add(token.start, kind(op));
    }
    result.add(input.len, 2);
    return result;
}

fn measure(io: std.Io, out: *std.Io.Writer, name: []const u8, op: []const u8, bytes: []const u8, comptime pass: fn ([]const u8) Result) !void {
    var iterations: usize = 1;
    while (true) {
        const start = std.Io.Clock.Timestamp.now(io, .awake);
        for (0..iterations) |_| std.mem.doNotOptimizeAway(pass(bytes));
        const elapsed = start.durationTo(std.Io.Clock.Timestamp.now(io, .awake)).raw.toNanoseconds();
        if (elapsed >= 50_000_000) break;
        iterations *= 2;
    }
    var samples: [15]u64 = undefined;
    var result: Result = .{};
    for (&samples) |*sample| {
        const start = std.Io.Clock.Timestamp.now(io, .awake);
        for (0..iterations) |_| {
            result = pass(bytes);
            std.mem.doNotOptimizeAway(result);
        }
        sample.* = @intCast(@divTrunc(start.durationTo(std.Io.Clock.Timestamp.now(io, .awake)).raw.toNanoseconds(), iterations));
    }
    try out.print("case={s} op={s} raw_samples={any}\n", .{ name, op, samples });
    std.mem.sort(u64, &samples, {}, std.sort.asc(u64));
    const median = samples[7];
    var devs: [15]u64 = undefined;
    for (samples, &devs) |s, *d| d.* = if (s > median) s - median else median - s;
    std.mem.sort(u64, &devs, {}, std.sort.asc(u64));
    try out.print("case={s} op={s} bytes={d} units={d} iterations={d} ns={d} mad={d} checksum={d}\n", .{
        name, op, bytes.len, result.count, iterations, median, devs[7], result.checksum,
    });
    try out.flush();
}

fn printStreams(out: *std.Io.Writer) !void {
    inline for (cases) |case| {
        try out.print("case={s} input=", .{case[0]});
        for (case[1]) |byte| try out.print("{x:0>2}", .{byte});
        try out.writeAll("\n");
        try out.print("case={s} stream=", .{case[0]});
        var it = lb.iterator(case[1]);
        while (it.next()) |boundary| {
            if (boundary.opportunity != .prohibited)
                try out.print("{d}:{s},", .{ boundary.offset, @tagName(boundary.opportunity) });
        }
        try out.writeAll("\n");
    }
    try out.flush();
}

pub fn main(init: std.process.Init) !void {
    var buf: [4096]u8 = undefined;
    var file = std.Io.File.stdout().writer(init.io, &buf);
    const out = &file.interface;
    const args = try init.minimal.args.toSlice(init.arena.allocator());
    if (args.len == 1 or (args.len == 2 and (std.mem.eql(u8, args[1], "--help") or std.mem.eql(u8, args[1], "-h")))) {
        try out.writeAll(
            \\Usage: line-break-diagnostic MODE
            \\
            \\Modes:
            \\  --bench  Run timing benchmarks
            \\  --dump   Print exact output records (bytes are hex-encoded)
            \\  --streams  Alias for --dump
            \\  --help, -h  Print this help
            \\
            \\No arguments prints help. Corpora are embedded at build time.
            \\
        );
        return out.flush();
    }
    if (args.len == 2 and (std.mem.eql(u8, args[1], "--dump") or std.mem.eql(u8, args[1], "--streams"))) return printStreams(out);
    if (args.len != 2 or !std.mem.eql(u8, args[1], "--bench")) return error.UnexpectedArgument;
    try out.writeAll("protocol=1 suite=linebreak peer=zunic engine=zunic-machine unicode=16.0.0 samples=15 calibration_ms=50 input=bytes consumption=offset_status_checksum_v1\n");
    inline for (cases) |case| {
        const bytes = case[1];
        var classifier = scalar.Classifier(false){};
        var pos: usize = 0;
        token_count = 0;
        while (pos < bytes.len) {
            const token = classifier.at(bytes, pos);
            tokens[token_count] = token;
            token_count += 1;
            pos = token.end;
        }
        if (!std.meta.eql(transitions(bytes), full(bytes))) return error.TransitionMismatch;
        try measure(init.io, out, case[0], "all_boundaries", bytes, allBoundaries);
        try measure(init.io, out, case[0], "opportunities_only", bytes, full);
        try measure(init.io, out, case[0], "decode_classify", bytes, classify);
        try measure(init.io, out, case[0], "preclassified_rules", bytes, transitions);
        try measure(init.io, out, case[0], "wrap_full", bytes, wrapFull);
        try measure(init.io, out, case[0], "wrap_24", bytes, wrapViewport);
    }
}

File: bench-vs-rust/rust-linebreak/diagnostic.rs
use std::{hint::black_box, time::Instant};
use unicode_linebreak::{break_property, linebreaks, BreakOpportunity};

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
struct Result {
    count: usize,
    checksum: u64,
}
fn full(bytes: &[u8]) -> Result {
    let text = std::str::from_utf8(black_box(bytes)).unwrap();
    let mut result = Result::default();
    for (offset, opportunity) in linebreaks(text) {
        result.count += 1;
        result.checksum = result.checksum.wrapping_mul(31).wrapping_add(offset as u64);
        let kind = match opportunity {
            BreakOpportunity::Allowed => 1,
            BreakOpportunity::Mandatory => 2,
        };
        result.checksum = result.checksum.wrapping_mul(31).wrapping_add(kind);
    }
    result
}

fn classify(bytes: &[u8]) -> Result {
    let text = std::str::from_utf8(black_box(bytes)).unwrap();
    let mut result = Result::default();
    for cp in text.chars() {
        result.count += 1;
        result.checksum = result
            .checksum
            .wrapping_add(break_property(cp as u32) as u64);
    }
    result
}

fn measure(name: &str, op: &str, text: &[u8], pass: impl Fn(&[u8]) -> Result) {
    let mut iterations = 1;
    loop {
        let start = Instant::now();
        for _ in 0..iterations {
            black_box(pass(text));
        }
        if start.elapsed().as_nanos() >= 50_000_000 {
            break;
        }
        iterations *= 2;
    }
    let mut samples = [0u128; 15];
    let mut result = Result::default();
    for sample in &mut samples {
        let start = Instant::now();
        for _ in 0..iterations {
            result = pass(text);
            black_box(result);
        }
        *sample = start.elapsed().as_nanos() / iterations;
    }
    println!("case={name} op={op} raw_samples={samples:?}");
    samples.sort_unstable();
    let median = samples[7];
    let mut devs = samples.map(|s| s.abs_diff(median));
    devs.sort_unstable();
    println!(
        "case={name} op={op} bytes={} units={} iterations={iterations} ns={median} mad={} checksum={}",
        text.len(),
        result.count,
        devs[7],
        result.checksum
    );
}

const HELP: &str = "Usage: diagnostic CORPUS_DIR MODE\n\nModes:\n  --bench  Run timing benchmarks\n  --dump   Print exact output records (bytes are hex-encoded)\n  --streams  Alias for --dump\n  --help, -h  Print this help\n\nNo arguments prints help. CORPUS_DIR is read before timing.\n";

fn main() {
    let cli: Vec<_> = std::env::args_os().collect();
    if cli.len() == 1 || (cli.len() == 2 && (cli[1] == "--help" || cli[1] == "-h")) {
        print!("{}", HELP);
        return;
    }
    if cli.len() != 3
        || !["--bench", "--dump", "--streams"]
            .iter()
            .any(|mode| cli[2] == *mode)
    {
        eprintln!("Invalid arguments.\n{HELP}");
        std::process::exit(2);
    }

    let directory = std::env::args().nth(1).expect("corpus directory");
    let streams = matches!(
        std::env::args().nth(2).as_deref(),
        Some("--dump" | "--streams")
    );
    let names = [
        "arabic",
        "hindi",
        "korean",
        "russian",
        "source_code",
        "english",
        "japanese",
        "mandarin",
    ];
    if streams {
        for name in names {
            let text = std::fs::read_to_string(format!("{directory}/{name}.txt")).unwrap();
            print!("case={name} input=");
            for byte in text.as_bytes() {
                print!("{byte:02x}");
            }
            println!();
            print!("case={name} stream=");
            for (offset, opportunity) in linebreaks(&text) {
                let kind = match opportunity {
                    BreakOpportunity::Allowed => "allowed",
                    BreakOpportunity::Mandatory => "mandatory",
                };
                print!("{offset}:{kind},");
            }
            println!();
        }
        return;
    }
    println!("protocol=1 suite=linebreak peer=rust engine=unicode-linebreak-0.1.5 unicode=15.0.0+SA samples=15 calibration_ms=50 input=bytes consumption=offset_status_checksum_v1");
    for name in names {
        let text = std::fs::read(format!("{directory}/{name}.txt")).unwrap();
        measure(name, "opportunities_only", &text, full);
        measure(name, "decode_classify", &text, classify);
    }
}
